docs/architecture/adr-015-ibc-packet-receiver.md

```diff
@@ -0,0 +1,201 @@
+}
+
+func (pvr ProofVerificationDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+  for _, msg := range tx.GetMsgs() {
+    var err error
+    switch msg := msg.(type) {
+    case client.MsgUpdateClient:
+      err = pvr.clientKeeper.UpdateClient(msg.ClientID, msg.Header)
+    case channel.MsgPacket:
+      err = pvr.channelKeeper.VerifyPacket(msg.Packet, msg.Proofs, msg.ProofHeight)
+    case chanel.MsgAcknowledgement:
+      err = pvr.channelKeeper.VerifyAcknowledgement(msg.Acknowledgement, msg.Proof, msg.ProofHeight)
+    case channel.MsgTimeoutPacket:
+      err = pvr.channelKeeper.VerifyTimeout(msg.Packet, msg.Proof, msg.ProofHeight, msg.NextSequenceRecv)
+    default:
+      continue
+    }
+
+    if err != nil {
+      return ctx, err.Result()
+    }
+  }
+  return next(ctx, tx, simulate)
+}
+```
+
+Where `MsgUpdateClient`, `MsgPacket`, `MsgAcknowledgement`, `MsgTimeoutPacket`
+are `sdk.Msg` types correspond to `handleUpdateClient`, `handleRecvPacket`,
+`handleAcknowledgementPacket`, `handleTimeoutPacket` of the routing module,
+respectively.
+
+The `ProofVerificationDecorator` will be inserted to the top level application.
+It should come right after the default sybil attack resistent layer from the
+current `auth.NewAnteHandler`:
+
+```go
+// add IBC ProofVerificationDecorator to the Chain of
+func NewAnteHandler(
+  ak keeper.AccountKeeper, supplyKeeper types.SupplyKeeper, ibcKeeper ibc.Keeper,
+  sigGasConsumer SignatureVerificationGasConsumer) sdk.AnteHandler {
+  return sdk.ChainAnteDecorators(
+    NewSetUpContextDecorator(), // outermost AnteDecorator. SetUpContext must be called first
+    ...
+    NewIncrementSequenceDecorator(ak),
+    ibcante.ProofVerificationDecorator(ibcKeeper.ClientKeeper, ibcKeeper.ChannelKeeper), // innermost AnteDecorator
+  )
+}
+```
+
+The Cosmos SDK will define the wrapper function `WriteAcknowledgement` under the
+ICS05 port keeper. The function will wrap packet handlers to automatically handle
+the acknowledgments.
+
+```go
+// Pseudocode
+type Ack struct {
+  Data []byte
+  Err sdk.Error
+}
+
+func (ack Ack) IsOK() bool {
+  return ack.Err.IsOK()
+}
+
+type PacketHandler func(sdk.Context, Packet) Ack
+
+func (k PortKeeper) WriteAcknowledgement(ctx sdk.Context, msg MsgPacket, h PacketHandler) sdk.Result {
+  // Cache context
+  cacheCtx, write := ctx.CacheContext()
+
+  // verification already done inside the antehandler
+  ack := h(cacheCtx, msg.Packet)
+  
+  // write the cache only if succedded
+  if ack.IsOK() {
+    write()
+  }
+  
+  res := ack.Err.Result()
+  // set the result to OK to persist the state change
+  res.Code = sdk.CodeOK
+  
+  // ackData will be stored as acknowledgement; []byte{} will be stored if not exists
+  if ack.Data == nil {
+    ack.Data = []byte{}
+  }
+  k.SetPacketAcknowledgement(ctx, msg.PortID, msg.ChannelID, msg.Sequence, ack.Data)
+
+  return res
+}
+```
+
+Example application-side usage:
```

Comment:
Thanks Joon, could you additionally write this handler with the specific implementation of ICS20 from `ibc-alpha`?

Reply:
i.e `CustomPacket ` should be the ICS20 transfer packet logic